### ISSUE TITLE
Support decomposed forms of strings

### ISSUE THREAD

**Issue by johnfactotum** [arty-name/locale-index-of#17]
It seems that it does not currently handle decomposed forms. Example:

```js
const collator = new Intl.Collator('fr', { sensitivity: 'base' })
indexOf(collator, 'caf\u0065\u0301 au lait, caf\u00e9 au lait', 'cafe au lait')
// expected: 0
// actual: 15
```

A simple solution would be to use `Intl.Segmenter` to align the strings:

```js
const indexOf_ = (segmenter, collator, string, substring) => {
    const substringLength = Array.from(segmenter.segment(substring)).length
    const segments = Array.from(segmenter.segment(string))

    for (let i = 0; i <= segments.length - substringLength; i++) {
        const potentialMatch = segments
            .slice(i, i + substringLength)
            .map(x => x.segment).join('')

        if (collator.compare(potentialMatch, substring) === 0) {
            return segments[i].index
        }
    }
    return -1
}

const segmenter = new Intl.Segmenter('fr', { granularity: 'grapheme' })
const collator = new Intl.Collator('fr', { sensitivity: 'base' })
indexOf_(segmenter, collator, 'caf\u0065\u0301 au lait, caf\u00e9 au lait', 'cafe au lait')
// => 0
```

**Comment by arty-name** [comment]
First I must say I don’t deal with intricacies of binary representation of Unicode much. 

So far, my understanding was that input should be normalized using `String.prototype.normalize()`, which by default produces composed versions of graphemes. I cannot judge how well the usage of decomposed versions is justified, but at least the support for producing them is also built-in into `normalize()`.

Thank you for offering a solution, but I am not sure that I can use it, since `Segmenter` is not supported in Gecko, and polyfilling it doesn’t look straightforward at all. I’m sorry about that. A proper solution should be native, but judging by the interest for this prollyfill, we should not hold our breath.

I’d recommend normalizing strings. If you must use decomposed forms, I think all the internals of the library are exported, so you can easily remix them for your use case, expanding on your suggestion:

 ```js
function localeIndexOf(string, substring, localesOrCollator, options) {
  const collator = getCollator(Intl, localesOrCollator, options);
  const segmenter = new Intl.Segmenter(collator.resolvedOptions().locale, { granularity: 'grapheme' });

  const substringLength = Array.from(segmenter.segment(substring)).length
  const segments = Array.from(segmenter.segment(string))

  for (let i = 0; i <= segments.length - substringLength; i++) {
    const potentialMatch = segments
      .slice(i, i + substringLength)
      .map(x => x.segment).join('')
   
    if (collator.compare(potentialMatch, substring) === 0) {
        return segments[i].index
    }
  }
  return -1
}

**Comment by arty-name** [comment]
Oh, I knew the name looked familiar! Kudos for the Foliate, it’s really great!

**Comment by arty-name** [comment]
I guess I can better see the limitations in the context of Foliate. Let me see what I can manage on the best effort basis.

**Comment by arty-name** [comment]
Could you try out [this version](https://github.com/arty-name/locale-index-of/blob/segmenter/index.js)?

**Comment by johnfactotum** [comment]
> So far, my understanding was that input should be normalized

The problem with normalization is that there's no way to map the result to the original string. In some cases this is okay, but sometimes one does need the precise index of the match in the original string.

Also I would suggest adding to the documentation a caveat that normalization might be required, if that's how this library is intended to be used.

> Could you try out this version?

Thanks! That looks very nice.

I guess one way to avoid the need of a segmenter would be to ignore all combining accents. That is, you build the potential match, adding one code point at a time, until it's as long as the needle, but any code point representing a combining accent isn't counted.

Also, I wonder if it's possible to use more sophisticated string searching algorithms (e.g. Boyer-Moore or its variations) to increase performance. (I understand that that might be outside the scope of this project. Just wondering, that's all.)

**Comment by arty-name** [comment]
> Also I would suggest adding to the documentation a caveat that normalization might be required, if that's how this library is intended to be used.

Well, that was an assumption about the best practice on my side.

> ignore all combining accents

That would require the ability to detect them, and generally seems like going in the direction of deep knowledge of internals, like does it only applies to combining characters? Meanwhile, the intended approach is "give me a comparator, and I will build my own indexOf". Your report shows that a tokenizer is needed too, but generally this library avoids being too smart. Which leads to an answer to the next question:

> use more sophisticated string searching algorithms

This is not something I would enjoy doing, and even though the prospect of my code powering a bit of Foliate warms me, it’s not worth it for me. As I was saying above, this is a stick to measure the interest for the feature, and 40 search results of `localeIndexOf` on GitHub show that it’s not really needed. Not enough for sophistication, in any case.

Reading some comments by smart people in https://github.com/tc39/ecma402/issues/506 is scary.

**Comment by arty-name** [comment]
I’ve pushed an indexing fix in the linked branch. If this version works properly for you, I could publish an update on npm.

**Comment by johnfactotum** [comment]
> That would require the ability to detect them, and generally seems like going in the direction of deep knowledge of internals, like does it only applies to combining characters?

I think since RegExp now supports [Unicode property escapes](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions/Unicode_Property_Escapes), one can use the Unicode property "Non-Spacing Mark", or `Mn`, to detect such characters. E.g.,

```js
/\p{Mn}/u.test('é') // '\u0065\u0301'
// => true
/\p{Mn}/u.test('é') // '\u00e9'
// => false
```

This isn't locale aware at all, so it probably won't work as expected in some cases, though it's probably good enough for searching purposes.

One can also take this as an alternative to using a collator. You can normalize the strings into decomposed forms, and then remove all combining accents using the RegExp. Then you can perform `indexOf`, `startsWith`, etc. without having to worry about diacritics.

> Your report shows that a tokenizer is needed too, but generally this library avoids being too smart.

Well, my understanding is that a grapheme segmenter would be the most "proper" way to align strings, as the definition of a "character" (to the end-user) can vary depending on the locale. It might be overkill in most cases, though.

Anyway, I think you're right. It's best to keep a library like this simple, and leave any advanced processing or normalization up to the user of the library. I'll close this, as the issue has been resolved. Thanks again!

**Comment by arty-name** [comment]
As a followup: I’ve published version 3.0.0 with the changes we’ve discussed
